shared/src/commonMain/kotlin/com/bunbeauty/shared/presentation/cafe_list/CafeListViewModel.kt

```diff
@@ -1,79 +1,96 @@
+                observeCart()
+                observeCafeList()
+            }
 
-    init {
-        observeCart()
-    }
+            is CafeList.Action.OnCafeClicked -> addEvent {
+                CafeList.Event.OpenCafeOptionsBottomSheet(uuid = action.cafeUuid)
+            }
 
-    private fun observeCart() {
-        sharedScope.launch(exceptionHandler) {
-            observeCartUseCase().collectLatest { cartTotalAndCount ->
-                mutableCafeItemListState.update { state ->
-                    state.copy(cartCostAndCount = cartTotalAndCount)
+            CafeList.Action.OnCartClicked -> {
+                addEvent {
+                    CafeList.Event.OpenConsumerCartProduct
                 }
             }
+
+            CafeList.Action.OnRefreshClicked -> {
+                observeCafeList()
+            }
         }
     }
 
-    fun observeCafeList() {
-        sharedScope.launch(exceptionHandler) {
-            observeCafeListJob?.cancel()
-            observeCafeListJob = observeCafeWithOpenStateListUseCase().onEach { cafeWithOpenStateList ->
-                mutableCafeItemListState.update { oldState ->
-                    oldState.copy(
-                        cafeList = cafeWithOpenStateList.map { cafeWithOpenState ->
-                            cafeWithOpenState.toCafeItem()
-                        },
-                        state = CafeListState.State.Success
-                    )
+    private fun observeCart() {
+        sharedScope.launchSafe(
+            block = {
+                observeCartUseCase().collectLatest { cartTotalAndCount ->
+                    setState {
+                        copy(
+                            cartCostAndCount = cartTotalAndCount
+                        )
+                    }
                 }
-            }.launchIn(sharedScope)
-        }
+            },
+            onError = {
+
+            },
+        )
     }
 
-    fun onCafeCardClicked(uuid: String) {
-        mutableCafeItemListState.update { oldState ->
-            oldState + CafeListState.Event.OpenCafeOptionsBottomSheet(
-                uuid = uuid
-            )
-        }
+    private fun observeCafeList() {
+        sharedScope.launchSafe(
+            block = {
+                observeCafeListJob?.cancel()
+                observeCafeListJob = observeCafeWithOpenStateListUseCase()
+                    .onEach { cafeWithOpenStateList ->
+                        setState {
+                            copy(
+                                cafeList = cafeWithOpenStateList.mapIndexed { index, cafeWithOpenState ->
+                                    cafeWithOpenState.toCafeItem(
+                                        isLast = index == cafeWithOpenStateList.lastIndex
+                                    )
+                                },
+                                isLoading = false
+                            )
+                        }
+                    }.launchIn(sharedScope)
```

Comment:
Already in `sharedScope` 
Lets just use `colect`

Reply:
we need job
How to get it with collect?

Reply:
need check